### DISCUSSION TITLE
Can't search for zip files

### DISCUSSION THREAD

**Discussion by xenos-github** [dnGrep/dnGrep#1182] [Q&A] [answered]
I'm searching folders for zip files. 
Using the pattern 

.*\\.zip$

I get no results. (I use this pattern because there are some files which embed the string "zip" within the filename.)
Using "docx", "txt", "step", "png" in place of "zip" all get hits. 
If I use just "zip" as my search pattern, I get one hit, which is a text file giving advice on how to unzip. This file contains "unzipping" in its name and ironically there is a zip file in the same folder. dnGrep appears to refuse to recognize "zip" as an extension. Indeed, 

.*\\...p$ 

and 

.*\\.\.ip$ 

fail too. 

"ep$" finds .step files, but "ip$" returns nothing. 

Any ideas? 

dnGrep: Version 4.2.59.0
Windows 11 Pro version 23H2

![Clipboard01](https://github.com/user-attachments/assets/dead0a49-9b4c-4030-abbd-0c694eac3c3d)

**Comment by xenos-github** [comment]
Correction: 

.*\\...p$

successfully locates .stp and other three-letter extensions ending with p, but no .zip files.

**Comment by doug24** [comment, marked as answer]
All of the archive extensions like zip are special cases - dnGrep sends them through a different path to treat them as a folder and read the contents.  I think what you are trying to do will work if you remove zip from the list of archive extensions in the Options dialog.  Then, zip files will be ordinary binary files, and you can find them, but you can't search inside them.

**Comment by xenos-github** [comment]
Thank you. I am wondering why I haven't hit this before. 
OK, I am not super desperate for dnGrep to be able to search inside archives, so I'm happy to wipe the archive extensions list. But this behaviour seems inconsistent with the option “List files using the file name pattern when the 'Search for' field is empty”.

**Reply by xenos-github** [reply]
There are many file types where the content is effectively immaterial; images, video, audio, so searching by name and filetype is common in this scenario. Likewise, I don't expect to be able to search archives as I don't know in advance the nature of their content, hence the ability to search archives by name, optionally without delving into their contents would be valuable to me. Cannot the “search in archives” option be used for this?

**Reply by xenos-github** [reply]
I can't get “search in archives” to do anything at all. 
If I clear the list of archive extensions, dnGrep still manages to list some text content from archive files (which _is_ useful), but when configured to recognize archives, it has not managed to open any of my archive files.

**Comment by doug24** [comment]
'Search in archives' only works if you tell dnGrep what file types are archives.  When you removed all the file extensions from the list of archives, dnGrep has no way to know which files are archives. dnGrep is not going to try to open every file as an archive just to see if it works.

You said originally that you wanted dnGrep to return a list of zip files. The way to do that is to remove 'zip' from the list of archives so dnGrep doesn't know that zip is an archive and will treat them as ordinary binary files. 

Once you have made your list of zip files, you need to put 'zip' back in the list of archives so dnGrep can again search inside zip files.  If it is not in the list, then the zip is just a blob of binary data - you may get something back from a search, but it is going to hit and miss.  It is the same as opening the zip file in Notepad - a few recognizable characters and a bunch of garbage.

dnGrep treats archive files differently because they are different than anything else in the filesystem.  Yes, they are binary files, but when you know how to open them, they look like directories.  The contain files, more directories, and can contain more archives.  Image, video and audio files do not share that property - they never contain something inside that dnGrep can search.

But dnGrep has to know what files to pass to the 7zip library to open them and return the contents as directories and files.  dnGrep does that by keeping a list of file extensions that are known to be archive extensions: zip, 7z, rar, jar, and so on.  When dnGrep sees one of these extensions, it goes down a different path in the code to expand it and return the contents.  To dnGrep the archive is another directory, not a file. It lists the contents of the archive, but not the archive itself.  As a result, you can't search for an archive as a file, just as you can't search for any other directory as if it were a file.  I think this a reasonable trade off.

Microsoft Office (docx, xlsx, pptx) and Open Office (odt) files are also archives.  But we don't put them in the list of archives because we have better ways to open those files with dnGrep plug-ins to extract the text data without the formatting.

**Reply by xenos-github** [reply]
Thank you with your patience. 
Perhaps I wasn't clear; I did re-instate the list of archive file extensions (zip rar jar etc) and still dnGrep failed to: 
a) locate any zip files, or 
b) open them. 
I was careful to use a search pattern that I knew would be found in both filenames and within archived files inside the zip archive.

**Comment by doug24** [comment]
I did misunderstand - I don't know why it would not locate or open any archive files.  You will have to show me all of your search settings - either a screen shot of the main window with filter area expanded or maybe it would be better to attach your dnGREP.Settings.dat file so I can see all the main screen settings and options settings.  Go to the About menu and choose "Application data and logs" to open the directory containing dnGREP.Settings.dat.

**Reply by xenos-github** [reply]
OK will do, but after the weekend. Thank you!

**Comment by xenos-github** [comment]
Well...
If I search using .* as the file pattern, it finds, opens and searches within zip archives. I have seen one .tgz file which it doesn't seem interested in - but how many levels of nested archive will it traverse? 
If I search using .*\.zip as the file pattern, it returns zero finds, both with and without a grep pattern for contents. I chose "matrix" knowing that there are both folders and files with this name. 
[Grep_Error_Log.zip](https://github.com/user-attachments/files/16759804/Grep_Error_Log.zip)

**Reply by doug24** [reply]
To search in archives, you need to check the "Search in archives" checkbox. Do not put archive file extensions in the "Patterns to match" - use this to put the file extensions you want to search in the archives.  dnGrep treats archives as directories, not files.  Putting `.*.\zip` in the Patterns to match will never match anything and filter out all files.

In the "Patterns to match", Use `.*` for all files, or limit it to txt, log, docx, or whatever.  You can specify multiple file types by separating the patterns with a semicolon.  Unless you are doing some special pattern to find a set of file names, it is probably easier to use asterisk patterns for 'Patterns to match' than regular expressions.

Try testing this with an archive other than tgz - tgz should work, with no limit on the number of nested archives - but I just tried it and there is a bug opening the unnamed archive at the root.  This used to work, I'll have to see what has changed. 

Unrelated, but I do not recommend using the option "Show results tree expanded" unless you know that there will be just a few small files in the results.

**Comment by xenos-github** [comment]
Hello Doug,

Splendid. Many thanks, and again, thank you for your patience. 

I'd say again, it's just a bit frustrating that I can't easily use dnGrep to find recently added archive files. I don't always want or need to search the contents of these files, just know where they are and maybe when they arrived. Switching off "Search in archives" would be the intuitive option, but your special handling of certain file extensions rules this out. . 

Chris Ramsden. 

Tuesday, August 27, 2024, 3:07:37 PM, you wrote:

To search in archives, you need to check the "Search in archives" checkbox. Do not put archive file extensions in the "Patterns to match" - use this to put the file extensions you want to search in the archives. dnGrep treats archives as directories, not files. Putting .*.\zip in the Patterns to match will never match anything and filter out all files.
In the "Patterns to match", Use .* for all files, or limit it to txt, log, docx, or whatever. You can specify multiple file types by separating the patterns with a semicolon. Unless you are doing some special pattern to find a set of file names, it is probably easier to use asterisk patterns for 'Patterns to match' than regular expressions.
Try testing this with an archive other than tgz - tgz should work, with no limit on the number of nested archives - but I just tried it and there is a bug opening the unnamed archive at the root. This used to work, I'll have to see what has changed.
Unrelated, but I do not recommend using the option "Show results tree expanded" unless you know that there will be just a few small files in the results.

**Comment by doug24** [comment]
Fixed in [4.2.84](https://github.com/dnGrep/dnGrep/releases/tag/v4.2.84.0)
